### ISSUE TITLE
Strictness information

### ISSUE THREAD

**Issue by ekmett** [well-typed/generics-sop#76]
I was using this to provide the generics interface for some code I was writing on stream recently.

However, I found I needed to know whether or not my fields were strict, which is something I can extract from the field selectors in GHC.Generics, but which doesn't seem to be present in the `generics-sop` metadata.

I wound up having to pull out all my `generics-sop` code out as a result, but I figured I should note here why in the hopes you might be able to address it in future releases.

**Comment by kosmikus** [comment]
Yes, I agree that this should be added.

**Comment by kosmikus** [comment]
I'm wondering how to best do this. We currently have:
```
data DatatypeInfo :: [[*]] -> * where
  -- Standard algebraic datatype
  ADT     :: ModuleName -> DatatypeName -> NP ConstructorInfo xss -> DatatypeInfo xss
  -- Newtype
  Newtype :: ModuleName -> DatatypeName -> ConstructorInfo '[x]   -> DatatypeInfo '[ '[x] ]

data ConstructorInfo :: [*] -> * where
  -- Normal constructor
  Constructor :: SListI xs => ConstructorName -> ConstructorInfo xs
  -- Infix constructor
  Infix :: ConstructorName -> Associativity -> Fixity -> ConstructorInfo '[ x, y ]
  -- Record constructor
  Record :: SListI xs => ConstructorName -> NP FieldInfo xs -> ConstructorInfo xs

data FieldInfo :: * -> * where
  FieldInfo :: FieldName -> FieldInfo a
  deriving (Show, Eq, Ord, Functor)
```
and a slightly simplified version of this for type-level metadata.

One option seems to be to add `POP StrictnessInfo xss` to the `ADT` constructor with
```
data StrictnessInfo :: * -> * where
  StrictnessInfo :: SourceStrictness -> SourceUnpackedness -> DecidedStrictness -> StrictnessInfo a
```
and `SourceStrictness`, `SourceUnpackedness`, `DecidedStrictness` being re-exported from `GHC.Generics`.

I first thought it should probably be added to `ConstructorInfo` or `FieldInfo`, but:
- `FieldInfo` is only used for record constructors, but strictness info is available and relevant for all fields;
- `ConstructorInfo` is a datatype with several constructors, but strictness info is relevant to all of them, so would have to be duplicated among these constructors, or put on the "outside";
- strictness information is not really applicable for `newtype`s; the GHC generics generates such info, but I think it's bogus information and we should be more precise if we can be.

**Comment by ekmett** [comment]
Having a separate `NP ConstructorInfo xss` and `POP StrictnessInfo xss` seems like it'd require a fair bit of manual zipping. What about storing an NP StrictnessInfo in Constructor and the other cases?

Also it seems your current form for `ConstructorInfo` is a bit problematic. 

You can have infix data constructors (that are even records!) with more than two fields.

```haskell
data Foo where 
    (:-) :: { foo :: Int, bar :: Int, baz :: Foo } -> Foo
    Nil :: Foo 
  deriving Show
```

typechecks perfectly well.

**Comment by edsko** [comment]
_stares fascinated_ but how does one use such an infix constructor?

**Comment by ekmett** [comment]
`(.)` is an infix operator that takes 3 arguments and we use it all the time!

Nothing stops you from working with `(:-)` with _other_ operators. I made it up on the spot, but let's see how it works with `(<<<)` from `Control.Category`.

```
ghci> 1 :- 2 <<< 3 :- 4 <<< 5 :- 6 $ Nil
(:-) {foo = 1, bar = 2, baz = (:-) {foo = 3, bar = 4, baz = (:-) {foo = 5, bar = 6, baz = Nil}}}
```

Then once the list of pairs is built up you can use record syntax to access it. Is this wise? Maybe not, but it is legal Haskell you can't talk about in the current API.

You can also build that up by using `(:-) { foo  = 10, bar = 12, baz = Nil }` syntax.

**Comment by ekmett** [comment]
I suppose they could just be collapsed into the other two cases.

**Comment by kosmikus** [comment]
So I guess we get something like
```
data ConstructorInfo :: [*] -> * where
  Constructor ::
        SListI xs
     => ConstructorName
     -> Fixity
     -> Maybe (NP FieldInfo xs) -- ^ for records
     -> NP StrictnessInfo xs
     -> Constructor xs
```
with just a single case, where `Fixity` is re-exported from `GHC.Generics` and not the `Fixity` type synonym we currently have; and `StrictnessInfo` is as proposed above.

**Comment by kosmikus** [comment]
When we're making backwards-incompatible changes to metadata anyway, I'm also tempted to rename `ADT` to `Data`.

**Comment by ekmett** [comment]
That looks right to me, modulo `s/Constructor/ConstructorInfo` on the last line.

**Comment by kosmikus** [comment]
@ekmett I looked at the infix stuff again, and I think I remember now why we ended up modelling it like we did (although I should have documented this better). It seems that the infix info provided by GHC Generics is purely syntactic, i.e., it reflects whether the constructor has been defined in prefix or infix notation.

Consider
```
data A =
     (:-) Int Int
   | Int :-- Int
   | Bar Int Int
   deriving Generic
infixr 5 :-
infixr 5 :--
infixr 5 `Bar`
```

This leads to
```
type instance Rep A
  = D1
      ('MetaData "A" "Ghci2" "interactive" 'False)
      (C1
         ('MetaCons ":-" 'PrefixI 'False)
         (S1
            ('MetaSel
               'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
            (Rec0 Int)
          :*: S1
                ('MetaSel
                   'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
                (Rec0 Int))
       :+: (C1
              ('MetaCons ":--" ('InfixI 'RightAssociative 5) 'False)
              (S1
                 ('MetaSel
                    'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
                 (Rec0 Int)
               :*: S1
                     ('MetaSel
                        'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
                     (Rec0 Int))
            :+: C1
                  ('MetaCons "Bar" 'PrefixI 'False)
                  (S1
                     ('MetaSel
                        'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
                     (Rec0 Int)
                   :*: S1
                         ('MetaSel
                            'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
                         (Rec0 Int))))
```
As you can see, only one of the three constructors is listed as infix, the other two as prefix. One can of course argue that this is wrong, but it seems difficult for us to do anything different, and then I think it is true that if `Infix` refers to constructors defined *syntactically* in infix position, it means these cannot be records, and cannot have anything else but two arguments.

**Comment by kosmikus** [comment]
If we keep the current structure of `ConstructorInfo`, the question returns to whether we should have a `POP StrictnessInfo` at datatype level, or rather an `NP StrictnessInfo` for each of the three cases `Constructor`, `Infix` and `Record`. If I understood Edward's earlier comments correctly, he'd prefer the latter. I don't have very strong feelings about this. Once the info is there, it's relatively easy to rearrange it. We already repeat the constructor name, so we can also repeat the strictness info.

**Comment by ekmett** [comment]
Note: You still get "strictness" information on the newtype today, it's just telling you the wrong thing.

**Comment by kosmikus** [comment]
@ekmett Do you mean in GHC.Generics? Yes, but that doesn't mean we have to do the same.

**Comment by kosmikus** [comment]
I've implemented the option of creating the `POP` here: 257e95facf3966c9df74da7bdebebcae40a89b99

This is not yet merged.

In any case, it does produce the necessary information both via the TH and GGP routes. So whatever representation we ultimately go for, I expect much of this code can then be reused.

Another option that would still be precise but distribute the strictness info over the constructors is to parameterize the `ConstructorInfo` datatype and use one variant, not carrying strictness info, in the `newtype` case, and the other, containing strictness info, in the `data` case. Would that be better?

**Comment by kosmikus** [comment]
See also #87.

**Comment by kosmikus** [comment]
Fixed via #87.
